test/rendering/cases/layer-vector-decluttering-extended/main.js

```diff
@@ -0,0 +1,188 @@
+import Style from '../../../../src/ol/style/Style.js';
+import Text from '../../../../src/ol/style/Text.js';
+import VectorLayer from '../../../../src/ol/layer/Vector.js';
+import VectorSource from '../../../../src/ol/source/Vector.js';
+import View from '../../../../src/ol/View.js';
+
+const center = [1825927.7316762917, 6143091.089223046];
+const map = new Map({
+  pixelRatio: 1,
+  target: 'map',
+  view: new View({
+    center: center,
+    zoom: 13.1,
+  }),
+});
+
+const sourceBlue = new VectorSource();
+sourceBlue.addFeatures([
+  new Feature({
+    geometry: new Point([center[0] + 1000 + 540, center[1] + 900 - 600]),
+    text: 'top-blue',
+  }),
+]);
+// circles are always drawn, but serve as obstacles,
+// texts are decluttered against each other and the circles
+map.addLayer(
+  new VectorLayer({
+    zIndex: 4,
+    declutter: true,
+    source: sourceBlue,
+    style: function (feature) {
+      return new Style({
+        image: new CircleStyle({
+          radius: 10,
+          stroke: new Stroke({
+            color: 'blue',
+            width: 8,
+          }),
+          declutterMode: 'declutter',
+        }),
+        text: new Text({
+          text: feature.get('text'),
+          font: 'italic bold 18px Ubuntu',
+          textBaseline: 'bottom',
+          offsetY: -15,
+        }),
+      });
+    },
+  })
+);
+
+const sourceRed = new VectorSource();
+sourceRed.addFeatures([
+  new Feature({
+    geometry: new Point([center[0] + 1000, center[1] + 1000 - 200]),
+    text: 'c-red',
+  }),
+  new Feature({
+    geometry: new Point([center[0] + 1000 - 540, center[1] + 1000]),
+    text: 'w-red',
+  }),
+  new Feature({
+    geometry: new Point([center[0] + 1000 + 540, center[1] + 1000 - 400]),
+    text: 'e-red',
+  }),
+]);
+// circles are always drawn, but serve as obstacles,
+// texts are decluttered against each other and the circles
+map.addLayer(
+  new VectorLayer({
+    zIndex: 3,
+    declutter: true,
+    source: sourceRed,
+    style: function (feature) {
+      return new Style({
+        image: new CircleStyle({
+          radius: 10,
+          stroke: new Stroke({
+            color: 'red',
+            width: 8,
+          }),
+          declutterMode: 'obstacle',
+        }),
+        text: new Text({
+          text: feature.get('text'),
+          font: 'italic bold 18px Ubuntu',
+          textBaseline: 'bottom',
+          offsetY: -15,
+        }),
+      });
+    },
+  })
+);
+
+const sourceOrange = new VectorSource();
+sourceOrange.addFeatures([
+  new Feature({
+    geometry: new Point([center[0], center[1]]),
+    text: 'c-orange',
+  }),
+  new Feature({
+    geometry: new Point([center[0] - 540, center[1]]),
+    text: 'w-orange',
+  }),
+  new Feature({
+    geometry: new Point([center[0] + 540, center[1]]),
+    text: 'e-orange',
+  }),
+]);
+// circles are always drawn,
+// texts are decluttered against each other and the layer 1 circles/texts
```

Comment:
For readability, please refer to the color instead of the layer number everywhere.

Reply:
comments improved.